Product Price block: add support to the Single Product Template

gigitux opened the pull request:
This PR added support to the Product Price block to Single Product Template.

Fixes https://github.com/woocommerce/woocommerce-blocks/issues/8479


On the implementation level, the main issue is that the atomic blocks support several contexts: 

- Inside `All Products Block` -> `withProductDataContext` HOC
- Inside `Products Block` -> Gutenberg Context
- Inside `Single Product Template` -> Gutenberg Context
- Without any parent -> `WithSelector` and `withProductDataContext` HOCs

#### withProductDataContext
The HOC withProductDataContext servers pass the select product object to the block.

#### With Selector
The HOC withSelector, when the `ProductDataContext` isn't defined, renders a modal for selecting the product and setting the product id in the context. In this way, the block can fetch the right data.

We added the support to different contexts at different times, so the code isn't in the best shape, which influenced the work for this PR.


It was necessary to create a component identical to the ´WithSelector` HOC. There is no need to show the selector product modal when the block is inside the Single Product template. Using a component, we can easily distinguish the two cases.

In any case, we should refactor all the atomic blocks and create a common logic easily to check and with great type safety: should we put it as a cooldown task? 

![image](https://user-images.githubusercontent.com/4463174/222519855-047667d0-61d2-4a0d-b7cc-ae5d131c076d.png)


### Testing
#### User Facing Testing

1. Switch to the TT3 theme.
2. Go to the Site Editor, and edit the Single Product Template.
3. Add the Product Price. It should have a price of 50.00 with the currency set in WC options. Save.
4. On the front end, click a product and check that the block shows the right price.

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Changelog

> Product Price block: add support to the Single Product Template.

gigitux committed:
Product Price Block: Add support Single Product Template

thealexandrelara commented on assets/js/atomic/blocks/product-elements/price/block.tsx:
```diff
@@ -86,12 +90,20 @@ export const Block = ( props: Props ): JSX.Element | null => {
+			regularPrice={
+				isDescendentOfSingleProductTemplate
+					? pricePreview
+					: prices.price
```
I'm wondering if since the lines 94-96 are using the same logic as these lines, if it would be better to store the logic in a variable/function. Just a precaution to prevent from someone from changing only one of them and forgot to change the other piece of the code in the future. Maybe something like this would work:
```
const price = isDescendentOfSingleProductTemplate
					? pricePreview
					: prices.price
...
return (
  <ProductPrice
    ...
    price={price}
    ...
    regularPrice={price}
   />
)
```

What do you think?

gigitux replied:
Good point, but for the regular price props, it is necessary to use `.regular_price`. I updated the code!

thealexandrelara reviewed (approved):
Great work! I added a comment in the code to check your opinion about it, but I'm pre-approving it.

One thing that I noticed is that the font-size for the Product Price seems smaller than the original one (in the Classic template), just want to confirm if this is expected or not:
| Classic  | Current |
| ------------- | ------------- |
| <img width="422" alt="image" src="https://user-images.githubusercontent.com/20469356/222544816-bd4099b2-4579-45aa-a459-709460163d40.png">  | <img width="429" alt="image" src="https://user-images.githubusercontent.com/20469356/222544886-d45b4b35-364f-432e-ad52-5ede6ea9705a.png">  |

imanish003 commented on assets/js/atomic/blocks/product-elements/shared/config.tsx:
```diff
@@ -14,7 +14,9 @@ import save from '../save';
-const sharedConfig: Omit< BlockConfiguration, 'attributes' | 'title' > = {
+const sharedConfig: Omit< BlockConfiguration, 'attributes' | 'title' > & {
+	ancestor: string[];
+} = {
```
Hey @gigitux, Should we consider contributing `ancestor` to [`Block` type here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/39e3b74c8998c623766ab21f9de2ddd21f76658e/types/wordpress__blocks/index.d.ts#L150)?

gigitux replied:
Great catch! Fixed with a dedicated PR: #8624

imanish003 commented on assets/js/atomic/blocks/product-elements/price/index.tsx:
```diff
@@ -17,25 +17,32 @@ import {
-const blockConfig: CustomBlockConfiguration = {
-	...sharedConfig,
+const blockConfig = {
+	...configuration,
```
It looks like you have removed `ancestor` from blockConfig. Therefore, this block could be added anywhere.
I tried adding this block to a new post and got following error: 
![image](https://user-images.githubusercontent.com/16707866/222658964-0e3191ed-e67a-4d75-bee7-0232e45259c5.png)

Are you able to reproduce it? Am I missing something 🤷🏻‍♂️

gigitux replied:
I forgot to import `_`. Fixed via a3d1ba37730ba796060e7b9b7e40594434ca2e20

imanish003 commented on assets/js/atomic/blocks/product-elements/price/edit.tsx:
```diff
@@ -51,26 +52,80 @@ const PriceEdit = ( {
+						postId ===
+							'woocommerce/woocommerce//single-product' ) &&
+					! isDescendentOfQueryLoop,
+			};
```
Is there any reason, because of which, we are returning Object instead of value?

gigitux replied:
Good point! Fixed via 8ccb2b305b42d0b9c5156836743f9c306d939409

imanish003 commented on assets/js/atomic/blocks/product-elements/price/edit.tsx:
```diff
@@ -51,26 +52,80 @@ const PriceEdit = ( {
+				icon={ BLOCK_ICON }
+				label={ BLOCK_TITLE }
+				description={ __(
+					'Choose a product to display its SKU.',
```
I think instead of `SKU`, we have to use `Price` here? 
```suggestion
					'Choose a product to display its Price.',
```

gigitux replied:
Good point! Fixed via a3d1ba37730ba796060e7b9b7e40594434ca2e20

imanish003 commented on assets/js/atomic/blocks/product-elements/price/block.tsx:
```diff
@@ -101,4 +113,9 @@ export const Block = ( props: Props ): JSX.Element | null => {
 };
 
-export default withProductDataContext( Block );
+export default ( props: Props ) => {
+	if ( props.isDescendentOfSingleProductTemplate ) {
+		return <Block { ...props } />;
+	}
+	return withProductDataContext( Block )( props );
+};
```
It's just a minor suggestion. In PR description, you have explained that `atomic blocks support several contexts`. Maybe we can add similar comment here to explain why we are returning early here?
Because it was not obvious to me until I looked at the PR description. What do you think? 🙂

gigitux replied:
Good point! Fixed via 9d8a43a60dcb16bda85c5dde508bebc63c9b5f79

imanish003 reviewed (changes requested):
Hi @gigitux, Well done! Except for minor things, everything else looks great to me 🙌🏻

I have left some minor suggestions but I don't have strong opinion about all of those 🙂

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into 8479-product-price-block-add-support-to-the-single-product-template

gigitux committed:
fix missing import

gigitux committed:
add comment

gigitux committed:
return a value

gigitux committed:
improve comment

gigitux committed:
fix import

gigitux commented:
Thanks folks, for your great comments!

@imanish003 I addressed all your feedback! 💪 

> One thing that I noticed is that the font-size for the Product Price seems smaller than the original one (in the Classic template), just want to confirm if this is expected or not:

Yep, the goal is to use the Global Styles to change the dimension of the text. By default, we will set Global Styles in the default HTML template. So, it will be necessary to update #8515

thealexandrelara reviewed (approved):
Nice! Thank you for addressing all the raised points! LGTM! 🚀 

> Yep, the goal is to use the Global Styles to change the dimension of the text. By default, we will set Global Styles in the default HTML template. So, it will be necessary to update https://github.com/woocommerce/woocommerce-blocks/pull/8515

Cool, so after we merge this one I'll add it to the Default HTML Template

gigitux committed:
Merge branch 'trunk' into 8479-product-price-block-add-support-to-the-single-product-template

gigitux committed:
Merge branch 'trunk' into 8479-product-price-block-add-support-to-the-single-product-template

imanish003 commented on assets/js/atomic/blocks/product-elements/shared/product-selector.tsx:
```diff
@@ -0,0 +1,77 @@
+								onClick={ () => setIsEditing( true ) }
+							>
+								{ __(
+									'Switch product…',
```
May I know if there is any reason behind adding `...` for `Switch product…'` text?

gigitux replied:
I don't know. It is the same string that there is in `WithProductSelector` HOC https://github.com/woocommerce/woocommerce-blocks/blob/e4325c12dd3bca8bb236454e9af90ce6416f6940/assets/js/atomic/blocks/product-elements/shared/with-product-selector.js/#L79-L80

imanish003 reviewed (approved):
Thank you for making the changes according to my suggestions. The updated code looks good to me. However, I have a minor question that I'd like to clarify, but it isn't a blocker for merging this pull request. So, I'm pre-approving the PR now. 🚀🙌🏻

gigitux committed:
Merge branch 'trunk' into 8479-product-price-block-add-support-to-the-single-product-template
